Add socket factory support to udp_server.c

mpwarres opened the pull request:
This allows users of udp_server to customize socket creation and/or binding via a channel_arg.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

mpwarres commented:
Ping. ctiller--any estimate on when this can be reviewed, or suggestions for a different reviewer if you are overloaded?

ctiller commented on src/core/lib/iomgr/udp_server.h:
```diff
@@ -58,12 +58,11 @@ typedef void (*grpc_udp_server_orphan_cb)(grpc_exec_ctx *exec_ctx,
 
 /* Create a server, initially not bound to any ports */
-grpc_udp_server *grpc_udp_server_create(void);
+grpc_udp_server *grpc_udp_server_create(struct grpc_server *server);
```
Why does this need a server pointer?

This is mixing stack levels... definitely against this change.

mpwarres replied:
grpc_udp_server already received a grpc_server*, but previously it was passed in grpc_udp_server_start().  The change here moves it to grpc_udp_server_create() so that the grpc_server's channel_args are available to grpc_udp_server when it is creating and binding UDP sockets.  Access to the channel_args is needed since they are how the grpc_socket_factory is passed to the grpc_udp_server.

Are you objecting to the use of channel_args for passing the socket factory, or to the way in which this plumbs channel_args (earlier) through to the grpc_udp_server?  What would you suggest instead?  Thanks...

ctiller replied:
I'd like to expunge grpc_server from udp_server entirely (what if I want a UDP socket without a grpc_server -- the use case for which is the reason this code is still in this repository)... I didn't realize it was already there, and I'd like to see it expunged quickly. 

It looks like it's there solely so it can be passed up in a read callback. A `void* user_data` field would be more appropriate there.

As for this usage, I think passing down channel args explicitly is entirely appropriate. You'll probably want to make a copy of them during construction (which is something we plan on making cheap... around EOQ3).

mpwarres replied:
Ok, I've expunged grpc_server from udp_server in https://github.com/grpc/grpc/pull/10021 , and updated this change to use channel_args directly.  PTAL again.  Thanks!

mpwarres commented:
Jenkins: test this please

mpwarres commented:
From offline conversation with @murgatroid99:
me:

> I noticed a couple failures when building with GRPC_UV defined--it looks like some issue with libpthread not being present.  AIUI, libuv is an alternative networking API to posix or windows sockets, and the socket factory abstraction added by my pull request only makes sense for posix sockets.  Should I scope src/core/lib/iomgr/socket_factory.[hc] to only expand to anything if GRPC_POSIX_SOCKET is #defined, and if so, does that mean these files should actually be named socket_factory_posix.[hc]?

@murgatroid99:

> There are a few things that should change there. First, the immediate cause of those compilation errors is that you are including sockaddr.h in socket_factory.h. Libuv can cause some import order problems, and the result of that is that sockaddr.h should be first in any list of included headers, and it should not be transitively included via other headers. Because of this, struct sockaddr should not be used in public or internal gRPC APIs. Instead, you should use the struct grpc_resolved_address, which is defined in resolve_address.h. The functions defined in sockaddr_utils.h allow for performing many operations on that type without accessing the underlying struct sockaddr.

> In addition, if the functionality is in fact specific to POSIX sockets, then the .c file should be enclosed in #ifdef GRPC_POSIX_SOCKET. I think we don't do that for .h files, under the assumption that platform-specific header files will only be included in corresponding platform-specific source files. Changing the name to socket_factory_posix.[ch] probably does make sense here. I can't comment on the names of the udp_server files; I wasn't involved in choosing them.

mpwarres commented:
Per @murgatroid99's suggestions, modified grpc_socket_factory to use grpc_resolved_address instead of sockaddr, and renamed socket_factory.[ch] to be socket_factory_posix.[ch].

@yang-g, could you take a look at this PR again, since it's changed somewhat from when @ctiller approved it?  Thanks...

mpwarres commented:
Thanks, @yang-g!

mpwarres commented:
Failures:
Basic Tests Mac: #9804
Basic Tests Windows: #8996
Tsan C++: #9150
